[css-anchor-position-1] Gate anchor-center behind CSS Anchor Positioning feature

tuankiet65 opened the pull request:
#### d9477327f62271df84745c05588d309783c3b4d2
<pre>
[css-anchor-position-1] Gate anchor-center behind CSS Anchor Positioning feature
<a href="https://rdar.apple.com/143097581">rdar://143097581</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=286112">https://bugs.webkit.org/show_bug.cgi?id=286112</a>

Reviewed by Tim Nguyen.

anchor-center should be gated behind the CSS Anchor Positioning feature,
but it wasn&apos;t. Fixes this by not parsing anchor-center if the feature is
not enabled.

* LayoutTests/fast/css/css-anchor-position/anchor-feature-disabled-expected.txt: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-feature-disabled.html: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
(WebCore::CSSPropertyParserHelpers::isSelfPositionKeyword):
(WebCore::CSSPropertyParserHelpers::consumeSelfPositionOverflowPosition):
(WebCore::CSSPropertyParserHelpers::consumeAlignSelf):
(WebCore::CSSPropertyParserHelpers::consumeJustifySelf):
(WebCore::CSSPropertyParserHelpers::consumeAlignItems):
(WebCore::CSSPropertyParserHelpers::consumeJustifyItems):
(WebCore::CSSPropertyParserHelpers::isSelfPositionOrLeftOrRightKeyword): Deleted.

Canonical link: <a href="https://commits.webkit.org/289170@main">https://commits.webkit.org/289170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/454a7cc49858a07204dfd5f57579f58557502d55

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85550 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36559 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87605 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66489 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24301 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4178 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46772 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4046 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31940 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35629 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74722 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32791 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92283 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9440 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75184 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13092 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73511 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74320 "Found 101 new API test failures: /TestWebKit:WebKit2UserMessageRoundTripTest.WKURLRequestRef, /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PageLoadState, /WebKitGTK/TestGeolocationManager:/webkit/WebKitGeolocationManager/current-position, /TestWebKit:WebKit.InjectedBundleBasic, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/create-navigation-data, /TestWebKit:WebKit.GeolocationBasic, /TestWebKit:WebKit2.ProvisionalURLAfterWillSendRequestCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/invalid-sequence, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/usermedia-enumeratedevices-permission-check ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18360 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18575 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17012 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4973 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12892 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18275 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12694 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16129 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14472 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
Can you add a layout test with `<!-- [ AnchorPositioningEnabled=false ] -->` to avoid regressing this?

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
```diff
@@ -75,14 +75,22 @@ static bool isContentPositionOrLeftOrRightKeyword(CSSValueID id)
+enum class AllowedSelfPositionKeywords {
+    LeftRight    = 1 << 0,
+    AnchorCenter = 1 << 1
+};
```
I think this might be a bit over engineered, can we just pass a bool for anchor center allowed?

tuankiet65 replied:
The code right now passes a function (`isSelfPositionKeyword` and `isSelfPositionOrLeftRightKeyword`) to check if a keyword is a self position keyword or not. If we add a bool parameter to those functions, then the functions that call those functions must also accept the bool and pass it along. ~~Also we'd have to make 4 check functions for `left`/`right` and `anchor-center` optional~~ (EDIT: it's either adding a bool to gate `anchor-center`, vs having 4 check functions). So I guess might as well use OptionSet.

tuankiet65 replied:
The naming could be better though, maybe OptionalSelfPositionKeywords

nt1m replied:
It's honestly fine either way, I didn't originally read the patch too much in detail but both probably are about same amount of code. Feel free to keep as is.

tuankiet65 committed:
[css-anchor-position-1] Gate anchor-center behind CSS Anchor Positioning feature
rdar://143097581
https://bugs.webkit.org/show_bug.cgi?id=286112

Reviewed by Tim Nguyen.

anchor-center should be gated behind the CSS Anchor Positioning feature,
but it wasn't. Fixes this by not parsing anchor-center if the feature is
not enabled.

* LayoutTests/fast/css/css-anchor-position/anchor-feature-disabled-expected.txt: Added.
* LayoutTests/fast/css/css-anchor-position/anchor-feature-disabled.html: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
(WebCore::CSSPropertyParserHelpers::isSelfPositionKeyword):
(WebCore::CSSPropertyParserHelpers::consumeSelfPositionOverflowPosition):
(WebCore::CSSPropertyParserHelpers::consumeAlignSelf):
(WebCore::CSSPropertyParserHelpers::consumeJustifySelf):
(WebCore::CSSPropertyParserHelpers::consumeAlignItems):
(WebCore::CSSPropertyParserHelpers::consumeJustifyItems):
(WebCore::CSSPropertyParserHelpers::isSelfPositionOrLeftOrRightKeyword): Deleted.

Canonical link: https://commits.webkit.org/289170@main
